Add gap junction connector

dwitvliet opened the pull request:
This adds a basic gap junction connector, which is compatible with the 3D Viewer and Connectivity widget. Gap junctions are currently ignored by all other widgets. The feature does not affect any current functions and does not clutter the compatible widgets (only adds one "show gap junctions" checkbox to the Connectivity widget). Gap junctions are added using the alt Key alone, as it did not previously have any functions (shift+alt and other combinations still work as previously).

We would like to use this feature to annotate and analyse gap junctions in _C. elegans_ serial electron microscopy data sets, where gap junctions are visible and essential for neuronal circuit functions. We hope that you agree with our additions, as it would help everyone in the _C. elegans_ community using CATMAID.

The feature has previously been suggested in [catmaid/CATMAID#333](https://github.com/catmaid/CATMAID/issues/333).

tomka commented:
Hi @dwitvliet, thanks a lot for your contribution! Being able to annotate Gap Junctions is certainly a nice improvement. Just out of curiosity, may I ask at what institute you are using CATMAID for *C. elegans* research?

So far I could only have a brief look at your code, but on a first glance everything seemed pretty nice. We will look at it in more detail (also to make sure performance of all back-end changes is fine) once it passes all tests: As you can see, our [tests](https://travis-ci.org/catmaid/CATMAID/builds/102223525) fail for your pull request. There are problems reported for both our front-end JSHint tests and our back-end tests.  The front-end part is easy to fix, there are only some missing semicolons:

```
django/applications/catmaid/static/js/widgets/connectivity.js: line 563, col 67, Missing semicolon.
django/applications/catmaid/static/js/widgets/overlay.js: line 1697, col 72, Missing semicolon.
django/applications/catmaid/static/js/widgets/overlay_node.js: line 1541, col 55, Missing semicolon.
```

With JSHint installed, you can run it like this from CATMAID's root directory:

```
jshint --config=.travis.jshintrc --exclude-path=.travis.jshintignore django/application
```

This should show no errors once you added the semicolons. For the back-end, you have to investigate a bit more it seems and maybe even modify the tests in `django/applications/catmaid/tests/test_common_apis.py`. For running the back-end tests you have to be in your virtualenv. Then you can run the tests with

```
manage.py test catmaid
```

Travis reports the following problems for the back-end:


```
======================================================================
ERROR: test_export_compact_skeleton (catmaid.tests.test_common_apis.ViewPageTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/catmaid/CATMAID/django/applications/catmaid/tests/test_common_apis.py", line 3518, in test_export_compact_skeleton
    self.assertItemsEqual(parsed_response[0], expected_response[0])
KeyError: 0
======================================================================
ERROR: test_node_list_without_active_node (catmaid.tests.test_common_apis.ViewPageTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/catmaid/CATMAID/django/applications/catmaid/tests/test_common_apis.py", line 3011, in test_node_list_without_active_node
    self.assertEqual(len(expected_t_result), len(parsed_response[0]))
KeyError: 0
======================================================================
FAIL: test_node_list_with_active_node (catmaid.tests.test_common_apis.ViewPageTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/catmaid/CATMAID/django/applications/catmaid/tests/test_common_apis.py", line 3055, in test_node_list_with_active_node
    self.assertEqual(4, len(parsed_response))
AssertionError: 4 != 2
======================================================================
FAIL: test_skeleton_connectivity (catmaid.tests.test_common_apis.ViewPageTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/catmaid/CATMAID/django/applications/catmaid/tests/test_common_apis.py", line 2203, in test_skeleton_connectivity
    self.assertEqual(expected_result, parsed_response)
AssertionError: {'outgoing_reviewers': [], 'outgoing': {'373': {'skids': {'235': [0, 0, 0, 0, 2] [truncated]... != {u'detail': u'Traceback (most recent call last):\n  File "/home/travis/virtualen [truncated]...
+ {u'detail': u'Traceback (most recent call last):\n  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/django/core/handlers/base.py", line 132, in get_response\n    response = wrapped_callback(request, *callback_args, **callback_kwargs)\n  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/django/utils/decorators.py", line 145, in inner\n    return func(*args, **kwargs)\n  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view\n    return view_func(*args, **kwargs)\n  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/django/views/generic/base.py", line 71, in view\n    return self.dispatch(request, *args, **kwargs)\n  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/rest_framework/views.py", line 466, in dispatch\n    response = self.handle_exception(exc)\n  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/rest_framework/views.py", line 463, in dispatch\n    response = handler(request, *args, **kwargs)\n  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/rest_framework/decorators.py", line 53, in handler\n    return func(*args, **kwargs)\n  File "/home/travis/build/catmaid/CATMAID/django/applications/catmaid/control/authentication.py", line 126, in inner_decorator\n    return f(request, *args, **kwargs)\n  File "/home/travis/build/catmaid/CATMAID/django/applications/catmaid/control/skeleton.py", line 918, in skeleton_info_raw\n    incoming, outgoing, gapjunctions, incoming_reviewers, outgoing_reviewers, gapjunctions_reviewers = _skeleton_info_raw(project_id, skeletons, op)\n  File "/home/travis/build/catmaid/CATMAID/django/applications/catmaid/control/skeleton.py", line 801, in _skeleton_info_raw\n    gapjunctions, gapjunctions_reviewers = _connected_skeletons(skeletons, op, relation_ids[\'gapjunction_with\'], relation_ids[\'gapjunction_with\'], relation_ids[\'model_of\'], cursor)\nKeyError: \'gapjunction_with\'\n',
+  u'error': u"'gapjunction_with'"}
- {'incoming': {'235': {'num_nodes': 28, 'skids': {'373': [0, 0, 0, 0, 2]}}},
-  'incoming_reviewers': [],
-  'outgoing': {'361': {'num_nodes': 9, 'skids': {'235': [0, 0, 0, 0, 1]}},
-               '373': {'num_nodes': 5, 'skids': {'235': [0, 0, 0, 0, 2]}}},
-  'outgoing_reviewers': []}
----------------------------------------------------------------------
```

It would be great if you could fix these problems and update the branch you used for this PR. Let us know if you need help to fix this.

aschampion commented:
Most of these tests are probably failing because the `gapjunction_with` relationship is missing in the test fixtures. This also breaks testing this branch on an existing project until re-running `catmaid.control.tracing.setup_tracing` to create the missing relation. @dwitvliet, you can add the `gapjunction_with` relation to the test fixtures in `django/applications/catmaid/fixtures/catmaid_testdata.json`, just inserting it near the `presynaptic_to` relations, etc. We'd also be happy to help out with any of this if you'd prefer. Just let us know.

As @tomka said, on first glance everything looks nice. One minor issue is that by default alt+click does not get sent to the browser in Gnome (Ubuntu), but we already have plans to add a radial menu when placing connectors to select types that would work around this.

Thanks for your contribution! Once these few small issues get sorted out, we can give the code a deeper look and hopefully merge soon.

dwitvliet commented:
Thank you, @tomka and @aschampion for the positive feedback and help! I've updated the frontend and backend to pass the tests. This should fix everything. Additionally, everywhere the gap junction relation is called, I have replaced `relation_map['gapjunction_with']` with `relation_map.get('gapjunction_with', -1)`, so the code still works if anyone hasn't re-run `catmaid.control.tracing.setup_tracing` to add the gap junction relation. 
@tomka To answer your question: We're using CATMAID in the Zhen Lab at Lunenfeld-Tanenbaum Research Institute, Toronto, Canada, in collaboration with multiple other groups, including, but not limited to, the Samuel and Lichtman groups at Center for Brain Science, Harvard, MA.

tomka commented:
Thanks for addressing these issues so quickly and for sharing information on your research project. Since the Alt key is also used for other actions in CATMAID, the use of Alt as modifier in your code shouldn't be a problem.

tomka commented:
Oh and now that I have actually tried your code and created gap junction connectors, I must say that this works really well. The display in the Connectivity Widget is us seamless and the 3D Viewer as well as  the Tracing Layer pick up the gap junctions just fine.

aschampion commented:
Thanks for the quick fixes. It works great, and we're impressed you managed to reuse many of the components with minimal changes (e.g., in the connectivity widget). There is a minor bug in the 3D viewer when switching between synapse coloring modes, but it's a result of the existing code needing to be refactored anyway, so I'll clean that up after merging.

While we haven't had much need for gap junctions in *Drosophila* yet, we know this has been on a lot of peoples' wishlists elsewhere for a long time, so thanks for this valuable contribution.

dwitvliet commented:
Thank you for testing it so quickly and integration the changes. CATMAID has really accelerated our research, and we are very glad to be able to contribute.

acardona commented:
@dwitvliet , thanks so much for this. One minor issue that I came across: when clicking on the gap junction count in the connectivity widget (a cell in the table), what opens is a table listing all the skeleton nodes etc involved, with one row per connection. These are titled "presynaptic node" and "postsynaptic node", when clearly here they should be named something else. Do you have a specific nomenclature for naming the partners on each side of a gap junction? In any case this issue can be addressed rather easily.

I noticed as well that you used "2" for the gap junction relation in the compact_arbor function in skeletonexport.py. While this works, a more general solution is needed for other types of relations (e.g. abutting). And a similar approach is necessary for the compact_arbor funtion. Derived from this is the imperative need to check whether anywhere in the CATMAID client it is assumed that "0" is pre and "else" is post. This would have to change to a switch statement with a proper "default" entry when the relation number does not match any of the expected ones.

dwitvliet commented:
@acardona Thank you for the feedback. I completely missed the issue with the connector table. I would be inclined to simply use "Neuron 1" and "Neuron 2". You could also do something more informative such as "Input neuron" and "Connected neuron", but that might create more confusion that help. I can correct that and submit another pull request if you prefer. I also noticed that the pre- and postsynaptic neurons seem to be switched in that table for synaptic connections (I am not sure if I introduced that bug or if it was there previously), so that needs a quick look at as well.

About your second concern with the compact-skeleton function, I just took a look a careful look, and there is indeed one case where gap junctions are counted as post in `exportwidget.html` because of an "if pre else post" statement. Additionally, there are more places abutting connectors are counted as post as well. I can take a careful look again and set up proper case statements where needed. I purposely didn't change the compact_arbor function to not interfere too much with the current code, so it only returns synaptic connectors, but I can add gap junctions to it as well and perform similar checks, if you prefer.

acardona commented:
@dwitvliet Would be wonderful if you could attempt a fix at the places that you identified where the new connector relations could be counted erroneously as either pre or post.
